File: LeetCode/516-longest-palindromic-subsequence/ex_516.cpp
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        if (s.empty())  return 0;
        int len = s.size();
        vector<vector<int>> dp(len, vector<int>(len, 1));
        for (int i = len - 1; i >= 0; i--) {
            for (int j = i + 1 ; j < len ; j++) {
                if (s[i] == s[j]) {
                    dp[i][j] = i +1 <= j - 1 ? dp[i+1][j-1] + 2 : 2;
                }
                else {
                    dp[i][j] = max(dp[i+1][j], dp[i][j-1]);
                }
            }
        }
        return dp[0][len-1];
    }
};

File: LeetCode/516-longest-palindromic-subsequence/ex_516.py
class Solution(object):
    def longestPalindromeSubseq(self, s):
        """
        :type s: str
        :rtype: int
        """
        if s == s[::-1]:    return len(s)
        n = len(s)
        dp = [[0] * n for _ in range(n)]
        for i in range(n):
            dp[i][i] = 1
        for j in xrange(1, len(s)):
            for i in reversed(xrange(0, j)):
                if s[i] == s[j]:
                    dp[i][j] = 2 + dp[i + 1][j - 1]
                else:
                    dp[i][j] = max(dp[i + 1][j], dp[i][j - 1])
        return dp[0][n-1]